php_phongo.c

```diff
@@ -1591,6 +1607,178 @@ void php_phongo_populate_default_ssl_ctx(php_stream_context *ctx, zval *driverOp
 #undef SET_STRING_CTX
 } /* }}} */
 
+bool php_phongo_apply_rp_options_to_client(mongoc_client_t *client, bson_t *options TSRMLS_DC) /* {{{ */
+{
+	bson_iter_t iter;
+	mongoc_read_prefs_t *new_rp;
+	const mongoc_read_prefs_t *old_rp;
+
+	if (!(old_rp = mongoc_client_get_read_prefs(client))) {
+		MONGOC_WARNING("Client does not have a read preference.");
+
+		return false;
+	}
+
+	/* Return early if there are no options to apply */
+	if (bson_empty0(options)) {
+		return true;
+	}
+
+	if (!bson_iter_init_find_case(&iter, options, "slaveok") &&
+	    !bson_iter_init_find_case(&iter, options, "readpreference") &&
+	    !bson_iter_init_find_case(&iter, options, "readpreferencetags")) {
+		return true;
+	}
+
+	new_rp = mongoc_read_prefs_copy(old_rp);
+
+	if (bson_iter_init_find_case(&iter, options, "slaveok") && BSON_ITER_HOLDS_BOOL(&iter)) {
+		mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_SECONDARY_PREFERRED);
+	}
+
+	if (bson_iter_init_find_case(&iter, options, "readpreference") && BSON_ITER_HOLDS_UTF8(&iter)) {
+		const char *str = bson_iter_utf8(&iter, NULL);
+
+		if (0 == strcasecmp("primary", str)) {
+			mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_PRIMARY);
+		} else if (0 == strcasecmp("primarypreferred", str)) {
+			mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_PRIMARY_PREFERRED);
+		} else if (0 == strcasecmp("secondary", str)) {
+			mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_SECONDARY);
+		} else if (0 == strcasecmp("secondarypreferred", str)) {
+			mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_SECONDARY_PREFERRED);
+		} else if (0 == strcasecmp("nearest", str)) {
+			mongoc_read_prefs_set_mode(new_rp, MONGOC_READ_NEAREST);
+		} else {
+			MONGOC_WARNING("Unsupported readPreference value [readPreference=%s].", str);
+		}
+	}
+
+	if (bson_iter_init_find_case(&iter, options, "readpreferencetags") && BSON_ITER_HOLDS_ARRAY(&iter)) {
+		bson_t tags;
+		uint32_t len;
+		const uint8_t *data;
+
+		bson_iter_array(&iter, &len, &data);
+
+		if (bson_init_static(&tags, data, len)) {
+			mongoc_read_prefs_set_tags(new_rp, &tags);
+		}
+	}
+
+	/* Warn on conflict, since read preference will be validated later */
+	if (mongoc_read_prefs_get_mode(new_rp) == MONGOC_READ_PRIMARY &&
+	    !bson_empty(mongoc_read_prefs_get_tags(new_rp))) {
+		MONGOC_WARNING("Primary read preference mode conflicts with tags.");
+	}
+
+	if (!mongoc_read_prefs_is_valid(new_rp)) {
+		mongoc_read_prefs_destroy(new_rp);
```

Comment:
The exception thrown by the caller will report that we couldn't apply read preferences (since we know that `php_phongo_apply_rp_options_to_client()` is what failed).